File: packages/gantt-4.2.7-trial/examples/Java CRUD demo/src/main/java/bryntum/gantt/request/handler/TaskSyncHandler.java
package bryntum.gantt.request.handler;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import bryntum.crud.exception.CrudException;
import bryntum.crud.request.SyncStoreRequestHandler;
import bryntum.gantt.dao.Gantt;
import bryntum.gantt.domain.Task;
import bryntum.gantt.domain.TaskSegment;

public class TaskSyncHandler extends SyncStoreRequestHandler<Task> {

    private final Gantt app;
    /**
     * Flag saying that segments changes are provided to the task being updated
     */
    private boolean isSegmentsProvided = false;

    public TaskSyncHandler(Gantt app, DateFormat dateFormat) {
        super(dateFormat);
        this.app = app;
    }

    protected TaskSegment getTaskSegment(Map<String, Object> changes) {
        Integer id = (Integer) changes.get("Id");

        TaskSegment record;

        if (id != null) {
            record = app.getTaskSegment(id);
            if (record == null) return null;
        } else {
            record = new TaskSegment();
        }

        if (changes.containsKey("Cls")) {
            record.setCls((String) changes.get("Cls"));
        }
        if (changes.containsKey("Duration")) {
            record.setDuration((Number) changes.get("Duration"));
        }
        if (changes.containsKey("DurationUnit")) {
            record.setDurationUnit((String) changes.get("DurationUnit"));
        }
        if (changes.containsKey("StartDate")) {
            try {
                record.setStartDate(strToDate((String) changes.get("StartDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }
        if (changes.containsKey("EndDate")) {
            try {
                record.setEndDate(strToDate((String) changes.get("EndDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }

        if (changes.containsKey(record.getPhantomIdField())) {
            record.setPhantomId((String) changes.get(record.getPhantomIdField()));
        }
        return record;
    }

    @SuppressWarnings("unchecked")
    @Override
    public Task getRecord(Map<String, Object> changes) {
        Integer id = (Integer) changes.get("Id");
        if (id == null) return null;

        Task record = app.getTask(id);
        if (record == null) return null;

        if (changes.containsKey("Color")) {
            record.setColor((String)changes.get("Color"));
        }
        if (changes.containsKey("ShowInTimeline")) {
            record.setShowInTimeline((Boolean)changes.get("ShowInTimeline"));
        }
        if (changes.containsKey("index")) {
            record.setIndex((Integer)changes.get("index"));
        }
        if (changes.containsKey("expanded")) {
            record.setExpanded((Boolean)changes.get("expanded"));
        }
        if (changes.containsKey("CalendarId")) {
            record.setCalendarId(changes.get("CalendarId"));
        }
        if (changes.containsKey("Cls")) {
            record.setCls((String) changes.get("Cls"));
        }
        if (changes.containsKey("Duration")) {
            record.setDuration((Number) changes.get("Duration"));
        }
        if (changes.containsKey("DurationUnit")) {
            record.setDurationUnit((String) changes.get("DurationUnit"));
        }
        if (changes.containsKey("StartDate")) {
            try {
                record.setStartDate(strToDate((String) changes.get("StartDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }
        if (changes.containsKey("EndDate")) {
            try {
                record.setEndDate(strToDate((String) changes.get("EndDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }

        if (changes.containsKey("Effort")) {
            record.setEffort((Number) changes.get("Effort"));
        }

        if (changes.containsKey("EffortUnit")) {
        	record.setEffortUnit((String) changes.get("EffortUnit"));
        }

        if (changes.containsKey("Note")) {
        	record.setNote((String) changes.get("Note"));
        }

        if (changes.containsKey("ConstraintType")) {
        	record.setConstraintType((String) changes.get("ConstraintType"));
        }

        if (changes.containsKey("ConstraintDate")) {
        	try {
				record.setConstraintDate(strToDate((String) changes.get("ConstraintDate")));
			} catch (Exception e) {
				// TODO Auto-generated catch block
				log.error(e);
			}
        }

        if (changes.containsKey("Draggable")) {
        	record.setDraggable((Boolean) changes.get("Draggable"));
        }

        if (changes.containsKey("Resizable")) {
        	record.setResizable((Boolean) changes.get("Resizable"));
        }

        if (changes.containsKey("Rollup")) {
        	record.setRollup((Boolean) changes.get("Rollup"));
        }

        if (changes.containsKey("ManuallyScheduled")) {
        	record.setManuallyScheduled((Boolean) changes.get("ManuallyScheduled"));
        }

        if (changes.containsKey("Name")) {
            record.setName((String) changes.get("Name"));
        }
        if (changes.containsKey("parentId")) {
            record.setParentId(changes.get("parentId"));
        }
        if (changes.containsKey("PercentDone")) {
            record.setPercentDone(changes.get("PercentDone"));
        }
        if (changes.containsKey("PhantomParentId")) {
            record.setPhantomParentId((String) changes.get("PhantomParentId"));
        }
        if (changes.containsKey("SchedulingMode")) {
            record.setSchedulingMode((String) changes.get("SchedulingMode"));
        }
        if (changes.containsKey("BaselinePercentDone")) {
            record.setBaselinePercentDone((Integer) changes.get("BaselinePercentDone"));
        }
        if (changes.containsKey("BaselineStartDate")) {
            try {
                record.setBaselineStartDate(strToDate((String) changes.get("BaselineStartDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }
        if (changes.containsKey("BaselineEndDate")) {
            try {
                record.setBaselineEndDate(strToDate((String) changes.get("BaselineEndDate")));
            } catch (Exception e) {
                log.error(e);
            }
        }
        if (changes.containsKey("Segments")) {
            isSegmentsProvided = true;

            List<Map<String, Object>> segmentsChanges = (List<Map<String, Object>>) changes.get("Segments");
            List<TaskSegment> segments = new ArrayList<TaskSegment>();

            // Segments might be null if we merge the task back
            if (segmentsChanges != null) {
                for (Map<String, Object> segmentChanges : segmentsChanges) {
                    segments.add(getTaskSegment(segmentChanges));
                }
            }

            record.setSegments(segments);
        } else {
            isSegmentsProvided = false;
        }

        if (changes.containsKey(record.getPhantomIdField())) {
            record.setPhantomId((String) changes.get(record.getPhantomIdField()));
        }

        return record;
    }

    protected Map<String, Object> prepareData(Task task) {
        // initialize returning hash
        Map<String, Object> result = new HashMap<String, Object>();

        String phantomTaskId = task.getPhantomParentId();
        if (task.getParentId() == null && phantomTaskId != null && !phantomTaskId.equalsIgnoreCase("root") && !phantomTaskId.equals("")) {
            Integer taskId = app.getTaskIdByPhantom(phantomTaskId);
            task.setParentId(taskId);
            result.put("parentId", taskId);
        }

        String phantomCalendarId = task.getPhantomCalendarId();
        if (task.getCalendarId() == null && phantomCalendarId != null && !phantomCalendarId.equals("")) {
            Integer calendarId = app.getCalendarIdByPhantom(phantomCalendarId);
            task.setCalendarId(calendarId);
            result.put("CalendarId", calendarId);
        }

        return result;
    }

    /**
     * Fills response with segments data.
     * Used to return identifiers of newly created segments.
     * @param response Whole task related response.
     * @param segments List of segments to return.
     */
    protected void addSegmentsResponse(Map<String, Object> response, List<TaskSegment> segments) {
        List<Map<String, Object>> segmentsResponse = new ArrayList<Map<String, Object>>();

        if (segments != null) {
            Map<String, Object> segmentResponse;

            for (TaskSegment segment : segments) {
                segmentResponse = new HashMap<String, Object>();
                segmentResponse.put("Id", segment.getId());
                segmentResponse.put(segment.getPhantomIdField(), segment.getPhantomId());
                segmentsResponse.add(segmentResponse);
            }

            response.put("Segments", segmentsResponse);
        } else {
            response.put("Segments", null);
        }
    }

    @Override
    public Map<String, Object> add(Task task) throws CrudException {
        Map<String, Object> result = prepareData(task);

        app.saveTask(task);

        // for a new task we always return segments response
        addSegmentsResponse(result, task.getSegments());

        return result;
    }

    @Override
    public Map<String, Object> update(Task task, Map<String, Object> changes) throws CrudException {
        Map<String, Object> result = prepareData(task);

        app.saveTask(task);

        // if segments changes were provided for the task being updated
        if (isSegmentsProvided) {
            addSegmentsResponse(result, task.getSegments());
        }

        return result;
    }

    @Override
    public Map<String, Object> remove(Task task) throws CrudException {
        Map<String, Object> result = new HashMap<String, Object>();
        app.removeTask(task);
        return result;
    }
}

File: packages/gantt-pro-4.2.7-trial/examples/PHP demo/php/Bryntum/Gantt/TaskSyncHandler.php
<?php

namespace Bryntum\Gantt;

class TaskSyncHandler extends \Bryntum\CRUD\SyncHandler
{

    protected $phantomIdField = 'PhantomId';
    private $gantt;
    protected $areSegmentsProvided = false;

    public function __construct(&$gantt)
    {
        $this->gantt = &$gantt;
        $this->phantomIdMap = &$gantt->phantomIdMap['tasks'];
        $this->segmentsPhantomIdMap = &$gantt->phantomIdMap['segments'];
    }

    protected function prepareTask(&$data)
    {
        // initialize returning hash
        $result = array();

        $this->areSegmentsProvided = array_key_exists('Segments', $data);

        $data['leaf'] = @$data['leaf'] ? 1 : 0;

        // Process datetime fields (turns field values to server timezone)

        if (isset($data['ConstraintDate'])) {
            $data['ConstraintDate'] = date('Y-m-d H:i:s', strtotime($data['ConstraintDate']));
        }

        if (isset($data['StartDate'])) {
            $data['StartDate'] = date('Y-m-d H:i:s', strtotime($data['StartDate']));
        }

        if (isset($data['EndDate'])) {
            $data['EndDate'] = date('Y-m-d H:i:s', strtotime($data['EndDate']));
        }


        if (isset($data['Duration'])) {
            $data['Duration'] = floatval($data['Duration']);
        }

        if (isset($data['Effort'])) {
            $data['Effort'] = floatval($data['Effort']);
        }

        if (isset($data['index'])) {
            $data['idx'] = intval($data['index']);
            unset($data['index']);
        }

        if (isset($data['expanded'])) {
            $data['expanded'] = intval($data['expanded']);
        }

        if (isset($data['ManuallyScheduled'])) {
            $data['ManuallyScheduled'] = intval($data['ManuallyScheduled']);
        }

        if (isset($data['Draggable'])) {
            $data['Draggable'] = intval($data['Draggable']);
        }

        if (isset($data['Resizable'])) {
            // Resizable is a string that might restrict direction allowed to resize to
            // but in this demo we support only true/false set of values
            $data['Resizable'] = intval($data['Resizable'] === '' || $data['Resizable']);
        }

        if (isset($data['Rollup'])) {
            $data['Rollup'] = intval($data['Rollup']);
        }

        if (isset($data['ShowInTimeline'])) {
            $data['ShowInTimeline'] = intval($data['ShowInTimeline']);
        }

        if (isset($data['parentId']) && (strtolower($data['parentId']) == 'root')) {
            $data['parentId'] = null;
        }

        // get newly created task Id if this is a reference to a phantom parent task
        if (isset($this->phantomIdMap[@$data['PhantomParentId']])) {
            // use & return actual Id
            $data['parentId'] = $result['parentId'] = $this->phantomIdMap[$data['PhantomParentId']];
        }

        $phantomCals = &$this->gantt->phantomIdMap['calendars'];
        // get newly created calendar Id if this is a reference to a phantom parent task
        if (isset($phantomCals[@$data['CalendarId']])) {
            // use & return actual Id
            $data['CalendarId'] = $result['CalendarId'] = $phantomCals[$data['CalendarId']];
        }

        if (isset($data['CalendarId'])) {
            $data['CalendarId'] = $data['CalendarId'] ? intval($data['CalendarId']) : null;
        }

        return $result;
    }

    protected function appendSegmentsResponse(&$response, $segments)
    {
        if (count($segments)) {
            $segmentsResponse = array();

            foreach ($segments as $segment) {
                $segmentsResponse[] = array('PhantomId' => $segment[$this->phantomIdField], 'Id' => $segment['Id']);
            }

            $response['Segments'] = $segments;
        }
    }

    public function add(&$task)
    {
        $response = $this->prepareTask($task);
        $this->gantt->saveTask($task);

        if ($this->areSegmentsProvided) {
            $this->appendSegmentsResponse($response, $task['Segments']);
        }

        return $response;
    }

    public function update(&$task)
    {
        $response = $this->prepareTask($task);
        $this->gantt->saveTask($task);

        if ($this->areSegmentsProvided) {
            $this->appendSegmentsResponse($response, $task['Segments']);
        }

        return $response;
    }

    public function remove($task)
    {
        $response = array();
        $this->gantt->removeTask($task);
        return $response;
    }

    protected function onRecordAdded($record, &$recordResponse, $phantomId)
    {
        parent::onRecordAdded($record, $recordResponse, $phantomId);

        // let's keep phantom Id to real Id mapping
        $this->phantomIdMap[$phantomId] = $recordResponse['Id'];

        return $recordResponse;
    }
}
